### PR TITLE
Laboratory work #4, Aleksandra Doroshina - 23FPL1

### PR DESCRIPTION
**Pull request by DoroshinaAleksandra**

### CODE REVIEW

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -26,6 +27,10 @@ def get_paragraphs(text: str) -> list[str]:
     Returns:
         list[str]: Paragraphs from document.
     """
+    if not text:
+        raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:31]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -36,10 +41,13 @@ class BM25Vectorizer(Vectorizer):
     _corpus: TokenizedCorpus
     _avg_doc_len: float
 
-    def __init__(self) -> None:
+    def __init__(self, corpus: TokenizedCorpus) -> None:
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py:44]: I think you should not change the interface

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
Comment (DoroshinaAleksandra) [reply lab_4_retrieval_w_clustering/main.py:32]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (DoroshinaAleksandra) [reply lab_4_retrieval_w_clustering/main.py]: fixed

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -170,6 +239,8 @@ def __init__(self, db: DocumentVectorDB) -> None:
         Args:
             db (DocumentVectorDB): Object of DocumentVectorDB class.
         """
+        self._db = db
+        super().__init__(self._db.get_vectorizer(), self._db.get_tokenizer())
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: better to call super as soon as possible.

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (DoroshinaAleksandra) [reply lab_4_retrieval_w_clustering/main.py]: fixed

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -40,6 +46,9 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        self._corpus = []
+        self._avg_doc_len = -1
+        super().__init__(self._corpus)
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Should be the first line in this method

File: lab_4_retrieval_w_clustering/main.py [hunk 5]
```diff
@@ -81,6 +100,15 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not tokenized_document or not isinstance(tokenized_document, list):
+            raise ValueError('inappropriate type of input argument or input argument is empty')
+        bm25_vector = [0.0] * len(self._vocabulary)
+        bm25 = calculate_bm25(self._vocabulary, tokenized_document, self._idf_values,
+                              k1=1.5, b=0.75, avg_doc_len=self._avg_doc_len, doc_len=len(tokenized_document))
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Do not set arguments that are defaults in this function (`k1`, `b`).

File: lab_4_retrieval_w_clustering/main.py [hunk 6]
```diff
@@ -182,6 +254,29 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
...
+            raise ValueError('used method return None')
+        knn = self._calculate_knn(vectorized_query, document_vectors, n_neighbours)
+        if knn is None:
+            raise ValueError('used method return None')
+        indices = []
+        for pair in knn:
+            indices.append(pair[0])
+        documents = self._db.get_raw_documents(tuple(indices))
+        if not documents:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 4 above)
Comment (DoroshinaAleksandra) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 5 above)
Comment (DoroshinaAleksandra) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 6 above)
Comment (DoroshinaAleksandra) [reply lab_4_retrieval_w_clustering/main.py]: done

**Pull request closed without merging**